pkg/query-service/rules/thresholdRule.go

```diff
@@ -499,32 +514,105 @@ func (r *ThresholdRule) runChQuery(ctx context.Context, db clickhouse.Conn, quer
 	return result, nil
 }
 
+func (r *ThresholdRule) prepareBuilderQueries(ts time.Time) (map[string]string, error) {
+	params := r.prepareQueryRange(ts)
+	runQueries := metrics.PrepareBuilderMetricQueries(params, constants.SIGNOZ_TIMESERIES_TABLENAME)
+
+	return runQueries.Queries, runQueries.Err
+}
+
+func (r *ThresholdRule) prepareClickhouseQueries(ts time.Time) (map[string]string, error) {
+	queries := make(map[string]string)
+
+	if r.ruleCondition == nil {
+		return nil, fmt.Errorf("rule condition is empty")
+	}
+
+	if r.ruleCondition.QueryType() != qsmodel.CLICKHOUSE {
+		zap.S().Debugf("ruleid:", r.ID(), "\t msg: unsupported query type in prepareClickhouseQueries()")
+		return nil, fmt.Errorf("failed to prepare clickhouse queries")
+	}
+
+	params := r.prepareQueryRange(ts)
+
+	for name, chQuery := range r.ruleCondition.CompositeMetricQuery.ClickHouseQueries {
+		if chQuery.Disabled {
+			continue
+		}
+		tmpl := template.New("clickhouse-query")
+		tmpl, err := tmpl.Parse(chQuery.Query)
+		if err != nil {
+			zap.S().Errorf("ruleid:", r.ID(), "\t msg: failed to parse clickhouse query to populate vars", err)
+			r.SetHealth(HealthBad)
+			return nil, err
+		}
+		var query bytes.Buffer
+		err = tmpl.Execute(&query, map[string]string{
+			"$__from_ts": fmt.Sprintf("%d", params.Start),
+			"$__to_ts":   fmt.Sprintf("%d", params.End),
+		})
+		if err != nil {
+			zap.S().Errorf("ruleid:", r.ID(), "\t msg: failed to populate clickhouse query", err)
+			r.SetHealth(HealthBad)
+			return nil, err
+		}
+		zap.S().Debugf("ruleid:", r.ID(), "\t query:", query.String())
+		queries[name] = query.String()
+	}
+	return queries, nil
+}
+
 // query looks if alert condition is being
 // satisfied and returns the signals
 func (r *ThresholdRule) buildAndRunQuery(ctx context.Context, ts time.Time, ch clickhouse.Conn) (Vector, error) {
-	params := r.prepareQueryRange(ts)
+	if r.ruleCondition == nil || r.ruleCondition.CompositeMetricQuery == nil {
+		r.SetHealth(HealthBad)
+		return nil, fmt.Errorf("invalid rule condition")
+	}
 
-	runQueries := metrics.PrepareBuilderMetricQueries(params, constants.SIGNOZ_TIMESERIES_TABLENAME)
-	if runQueries.Err != nil {
-		return nil, fmt.Errorf("failed to prepare metric queries: %v", runQueries.Err)
+	// var to hold target query to be executed
+	queries := make(map[string]string)
+	var err error
+
+	// fetch the target query based on query type
+	if r.ruleCondition.QueryType() == qsmodel.QUERY_BUILDER {
+
+		queries, err = r.prepareBuilderQueries(ts)
+
+		if err != nil {
+			zap.S().Errorf("ruleid:", r.ID(), "\t msg: failed to prepare metric queries", zap.Error(err))
+			return nil, fmt.Errorf("failed to prepare metric queries")
+		}
+
```

Comment:
Unlike `CLICKHOUSE`, this doesn't set (not here and not in the callee function) the health to bad for non-nil errors? You could set the health in this function (or a top-level function) to make it easier for the reader to follow instead of having it in multiple places.

Reply:
Agree. I haven't really touched the health condition "bad" completely yet in the interest of time. As such it is right now not much helpful, but i was thinking we can start showing it up in the alert listing page.
this requires some more thinking, such as it would help in having a common logging function that prints vars.